skmorf/tree/_unsup_criterion.pyx

```diff
@@ -133,35 +160,181 @@ cdef class TwoMeans(Criterion):
+
+        mu_left = X[:pos].sum()
+        mu_right = X[pos:end].sum()
+
+        for k in range(pos):
+            left_s += (X[k] - mu_left)**2
+
+        for c in range(pos,end):
+            right_s += (X[c] - mu_right)**2
+
+        impurity = left_s + right_s
+        impurity /= self.weighted_n_node_samples
+
+        return impurity
 
     cdef void children_impurity(self, double* impurity_left,
                                 double* impurity_right) nogil:
-        # need to implement
+        # cdef double entropy_left = 0.0
+        # cdef double entropy_right = 0.0
+        # cdef double count_k
+        # cdef SIZE_t k
+        # cdef SIZE_t c
+
+        # for k in range(self.n_outputs):
+        #     for c in range(self.n_classes[k]):
+        #         count_k = self.sum_left[k, c]
+        #         if count_k > 0.0:
+        #             count_k /= self.weighted_n_left
+        #             entropy_left -= count_k * log(count_k)
+
+        #         count_k = self.sum_right[k, c]
+        #         if count_k > 0.0:
+        #             count_k /= self.weighted_n_right
+        #             entropy_right -= count_k * log(count_k)
+
+        # impurity_left[0] = entropy_left / self.n_outputs
+        # impurity_right[0] = entropy_right / self.n_outputs
+        
+        # use left_sum and right_sum to calculate impurity?
+        # or should this method take two separate Xs making up
+        # left and right child nodes?
         pass
 
     cdef void node_value(self, double* dest) nogil:
-        """Compute the node value of samples[start:end] and save it into dest.
+        r"""Compute the node value of samples[start:end] and save it into dest.
+
+        Node-Wise Feature Generation
+        URerF doesn't choose split points in the original feature space
+        It follows the random projection framework
+
+        \tilde{X}= A^T X'
+
+        where, A is p x d matrix distributed as f_A, where f_A is the 
+        projection distribution and d is the dimensionality of the 
+        projected space. A is generated by randomly sampling from 
+        {-1,+1} lpd times, then distributing these values uniformly 
+        at random in A. l parameter is used to control the sparsity of A
+        and is set to 1/20.
+
+        Each of the d rows \tilde{X}[i; :], i \in {1,2,...d} is then 
+        inspected for the best split point. The optimal split point and 
+        splitting dimension are chosen according to which point/dimension
+        pair minimizes the splitting criteria described in the following 
+        section
 
         Parameters
         ----------
         dest : double pointer
             The memory address which we will save the node value into.
         """
-        # need to implement
-        pass
+        cdef SIZE_t k
+
+        for k in range(self.n_outputs):
+            memcpy(dest, &self.sum_total[k, 0], self.n_classes[k] * sizeof(double))
+            dest += self.max_n_classes
```

Comment:
No notion of class. 

Possible tip: See how `node_value` is used in splitter and tree.